### PR TITLE
Adds UART RX IRQ Callback with onReceive()

### LINKED ISSUE #5472: HardwareSerial: No way to wait for data in queue.
### Hardware:
Board:							ESP32
Core Installation version:			framework-arduinoespressif32 3.10006.210326 (1.0.6)
IDE name:						Platform.io
Flash Frequency:					40Mhz
PSRAM enabled:                                    no
Upload Speed:					115200
Computer OS:                                        Windows 10

### Description:
I have worker task, which should sleep while no incoming data on uart and  wake on data and process it.
I found that HardwareSerial::available() and HardwareSerial::read() 
are non-blocking. So i have busyloop in task which is not good for CPU resource optimization.

### Sketch: 
```cpp

#include <Arduino.h>
#include <HardwareSerial.h>

void *uart_task(void *arg)
{
	size_t read;
	for (;;) {
		wait_for_data(&Serial2); // <- How to do this????

		if (!Serial2.available()) // This is non-blocking.
			continue;

		read = Serial2.read(buff, sizeof(buff)); // Also non-blocking.
		process_data(buff, read);
	}
}

```

### LINKED ISSUE #5620: Serial trigger 1 byte is possible????
### Hardware:
|||
|:---|---|
|<B>Board</B>|ESP32 Dev Module|
|<B>Version/Date</B>|2.0.0|
|<B>IDE name</B>|Arduino IDE|
|<B>Flash Frequency</B>|80Mhz|
|<B>PSRAM enabled</B>|no|
|<B>Upload Speed</B>|115200|
|<B>Computer OS</B>|Windows 10|

### Description:
Hey Good afternoon,

I have a doubt, and it is possible to make a trigger for each byte received by Serial.
I'm using the following code but it doesn't.

```cpp


#include <Arduino.h>


#define PIN_NOT_RE 18   // LOW = Recived
#define PIN_DE 19  // HIGH = SEND

void setup() {
  // put your setup code here, to run once:

  Serial.begin(14400);
  Serial.setRxBufferSize(1);
  Serial.println(Serial.getTimeout()); // print the default value
  Serial.setTimeout(1);
  Serial.println(Serial.getTimeout()); // print the new value
  pinMode(4, OUTPUT);
  pinMode(PIN_NOT_RE, OUTPUT);
  pinMode(PIN_DE, OUTPUT);

  delay(5000);


  digitalWrite(4, HIGH);

  digitalWrite(PIN_NOT_RE, LOW);
  digitalWrite(PIN_DE, LOW);
}


void loop() {

}



void serialEvent() {

  if (Serial.available()) {
    digitalWrite(4, LOW);
    // get the new byte:
    char inChar = (char)Serial.read();
    digitalWrite(4, HIGH);
  }

}


```
Captura de ecrã 2021-09-01 155422

### LINKED ISSUE #5678: Arduino 2.0.0 - Question: wants to use UART Rx interrupt but still keeps HardwareSerial API to handles data from Rx FIFO
### Hardware:
Board:							ESP32 Dev Module
Core Installation version:			2.0.0
IDE name:						PlatformIO
Flash Frequency:					40Mhz
PSRAM enabled:                                      no
Upload Speed:					921600
Computer OS:                                          Mac OSX

### Description:
Currently, I got UART interrupt working fine. I have created new class to inherits `HardwareSerial` then overrides `begin()` to enables interrupt routine. As my understanding from [ESP32 docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/uart.html#using-interrupts), when registered interrupt isr, its override default isr handler. This means the Rx `HardwareSerial` that relies on IDF's API (eg. `read`, `available`, etc.) will be unusable, need to handles from isr by user. So, I have override HardwareSerial functions to points to a new create buffer, queue that handles the data from isr (likes an old implementation in 1.0.6).

I have take a look at `uart-event` but if i want to unpack my serial packet per byte, i think isr still the way to go. correct me if I misunderstanding.

Is there any other solution to workaround this if I still want to use HardwareSerial functions (some libraries that based on default HardwareSerial functions) while using UART interrupt.

Cheers,
MomePP

### LINKED ISSUE #6102: how to use esp32-hal in arduino,example esp32-hal-uart.h  for serial interrupt  in arduino?
### Board

esp32

### Device Description

esp32

### Hardware Configuration

esp32

### Version

latest master

### IDE Name

platformio

### Operating System

win10

### Flash frequency

40

### PSRAM enabled

no

### Upload speed

9600

### Description

i want to use serial hardware interrupt in arduino,but i found the arduino api _serialEvent()_ is not really hardware serial interrupt.so,i fonud some in _.platformio\packages\framework-arduinoespressif32\cores\esp32_   ,**esp32-hal-uart.c** and **esp32-hal-uart.h**
i think this two file can help me. but i dont find any help document about this HAL file.

//example this function
uart_t* uartBegin(uint8_t uart_nr, uint32_t baudrate, uint32_t config, int8_t rxPin, int8_t txPin, uint16_t queueLen, bool inverted);

what uart_nr?   what config?  what queueLen?i dontknow rev data len. what inverted?

### Sketch

```cpp
//example this function
uart_t* uartBegin(uint8_t uart_nr, uint32_t baudrate, uint32_t config, int8_t rxPin, int8_t txPin, uint16_t queueLen, bool inverted);

what uart_nr?   what config?  what queueLen?i dontknow rev data len. what inverted?
```

### Debug Message

```plain
....
```

### Other Steps to Reproduce

_No response_

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by SuGlider**
## Summary
This PR adds a new feature to UART / Arduino Serial API.

Users have requested a feature related to UART ISR that allow a callback function to be executed as soon as data arrives to UART port. This PR adds ```Serial.onReceive()``` functionality to Arduino Hardware Serial.

Example:
``` dart
void UART_RX_IRQ() {
  uint16_t size = Serial.available();
  Serial.printf("Got %d bytes on Serial to read\n", size);
  while(Serial.available())  {
    Serial.write(Serial.read());
  }
  Serial.printf("\nSerial data processed!\n");
}

void setup() {
  Serial.begin(115200);
  Serial.onReceive(UART_RX_IRQ);
  Serial.println("Send data to UART0 in order to activate the RX callback");
}

void loop() {
  Serial.println("Sleeping for 10 seconds...");
  delay(10000);
}
```

Output of the example in the ESP32-C3 (```Hello World!``` sent to default Serial port using Arduino Serial Monitor):

``` dart
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd6100,len:0x420
load:0x403ce000,len:0x90c
load:0x403d0000,len:0x236c
SHA-256 comparison failed:
Calculated: ccb0d00bac7e84e1d90a12e4f75f4ab6c5f7e71bb209afd5819c4c9557a6db71
Expected: c9cf160580940ec7801c73b16423824e72ad12055c732e83ce66332240af42a7
Attempting to boot anyway...
entry 0x403ce000
Send data to UART0 in order to activate the RX callback
Sleeping for 10 seconds...
Sleeping for 10 seconds...
Got 13 bytes on Serial to read
Hello World!

Serial data processed!
Sleeping for 10 seconds...

````

## Impact
It includes a new feature to Arduino Hardware Serial.
It needs to be documented (@pedrominatel) 

## Related links

Initial PR requesting the feature
PR #4656
PR #3376 

Related Issues
Fixes #5678
Fixes #5620 
Fixes #5472
Fixes #6102

### CODE REVIEW

File: cores/esp32/esp32-hal-uart.c [hunk 1]
```diff
@@ -82,6 +84,65 @@ uint32_t _get_effective_baudrate(uint32_t baudrate)
...
+{
+    uart_t* uart = (uart_t *)args;
+    uart_event_t event;
+    for(;;) {
+        //Waiting for UART event.
+        if(xQueueReceive(uart->uart_event_queue, (void * )&event, (portTickType)portMAX_DELAY)) {
+            switch(event.type) {
+                //Event of UART receving data
+                case UART_DATA:
+                    if(uart->onReceive) uart->onReceive();
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-uart.c:109]: shouldn't there be a mutex lock here too? what if `onReceive` get's cleared right after it was checked?
TBH I probably would not have added locks to `uartOnReceive` either :) but it is in reality possible to cause an exception.

**Comment by me-no-dev** [comment]
I like it :) let's discuss my single comment and we can merge it.

**Comment by SuGlider** [comment]
Adds LOCKs to UART Event Callback Task.

File: cores/esp32/esp32-hal-uart.c (hunk 1 above)
Comment (SuGlider) [reply cores/esp32/esp32-hal-uart.c:110]: Done!

**Review by SuGlider — COMMENTED** [review]
Ready for merging.

**Comment by FredGlaum** [comment]
Hi I recently also experienced buffer overrun errors on high baudrates - I am glad to see there is solution in the works.  When can we expect to use the onReceive() event callback for the Arduino/PlatformIO  ESP32 development. 

I tried the latest dev and stable core releases and its not there.

**Comment by SuGlider** [comment]
> Hi I recently also experienced buffer overrun errors on high baudrates - I am glad to see there is solution in the works. When can we expect to use the onReceive() event callback for the Arduino/PlatformIO ESP32 development.
> 
> I tried the latest dev and stable core releases and its not there.

PlaformIO ESP32 still uses old Arduino Core version 1.0.6.
This patch depends on IDF and it will be available in Arduino Core version 2.0.3+.
You can try to manually setup your PlatformIO to use the latest master version to use this feature.

Maybe this link may help you:
https://community.platformio.org/t/use-different-version-of-arduino-framework-on-espressif-platform/23669

**Comment by gemu2015** [comment]
@SuGlider 
could you please verify this pr since it seems to break UART receive in some users of Tasmota.
under some circumstances receive stops and only reverts after a restart. this is fixed after reverting this pr.
since this option is only needed in rare cases i would also prefer to not enable it by default since it creates an extra queue and task wasting memory and cpu cycles.
probably only create queue and task when uartOnReceive is actually called ?

**Comment by SuGlider** [comment]
@gemu2015 
This PR has an issue fixed and merged to master in PR #6201
This shall fix the issue you report as "break UART receive".

**Comment by SuGlider** [comment]
> probably only create queue and task when uartOnReceive is actually called ?

This is not possible because of IDF UART driver setup. 
IDF is the base for all HAL layer of ESP32 Arduino in Arduino Core Version 2.0.0+.

**Comment by gemu2015** [comment]
> @gemu2015
> This PR has an issue fixed and merged to master in PR #6201
> This shall fix the issue you report as "break UART receive".

however we do not use HardwareSerial::onReceive() at all and get locks too ??

**Comment by SuGlider** [comment]
> however we do not use HardwareSerial::onReceive() at all and get locks too ??

As said it is fixed. You are talking about WIP - this is not released yet.

**Comment by Jason2866** [comment]
For completness tested with applying #6201 too. Still issues. Maybe it is related to #6326

**Comment by SuGlider** [comment]
@Jason2866 - Please check #6364 because it shall be the final version of HardwareSerial and esp32-hal-uart for the next Arduino Core Version 2.0.3.

I fixed a number of issues related to HardwareSerial overflows, locking issues, onReceive() and others.

**Comment by Jason2866** [comment]
@SuGlider thx, will do a build with. @gemu2015 when it is compiled i will drop you a note.

**Comment by dagi0627** [comment]
My esp32 (device1) connected to another MCU (device2) over UART2.  Device2 sends some data at every second then device1 recieveing it well and send back over UART0 for show on terminal. It is good working. But some in case device1 or esp32 send character "@" to device2 then device2 stop to send data at everysecond and send some information at once time . That is same following 
``` 1. Lorem Ipsum is simply dummy text of the printing and typesetting industry.
2. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
3. when an unknown printer took a galley of type 
4. and scrambled it to make a type specimen book. 
5. It has survived not only five centuries, 
6. but also the leap into electronic typesetting, 
7. remaining essentially unchanged. 
8. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, 
9. and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.```  
But it is not show on terminal. After few minute Device2 send data at everysecond then i show above information. Why it is Buffered data.

**Pull request merged**